// @wycats/agent-hooks/src/index.js
/**
 * @wycats/agent-hooks
 *
 * Cross-platform hook runtime for VS Code and Claude Code.
 *
 * Handles stdin/stdout boilerplate, tool name normalization,
 * and platform-appropriate output formatting. Hook scripts
 * become thin wrappers that express check/handle logic.
 */
import { readFileSync, appendFileSync, mkdirSync } from "node:fs";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";
const toolsPath = join(dirname(fileURLToPath(import.meta.url)), "..", "tools.json");
const toolsData = JSON.parse(readFileSync(toolsPath, "utf-8"));
// Build the flat map: platform tool name → canonical name
const TOOL_MAP = {};
for (const [, entry] of Object.entries(toolsData)) {
    if (typeof entry === "string")
        continue; // skip $comment
    for (const platformTools of Object.values(entry.platforms)) {
        for (const toolName of platformTools) {
            TOOL_MAP[toolName] = entry.canonical;
        }
    }
}
export function normalizeToolName(raw) {
    return TOOL_MAP[raw] ?? raw;
}
/** The raw tool normalization map for inspection/testing. */
export const toolMap = TOOL_MAP;
function readInput() {
    return JSON.parse(readFileSync("/dev/stdin", "utf-8"));
}
function detectContext(raw) {
    const str = (key) => {
        const v = raw[key];
        return typeof v === "string" ? v : undefined;
    };
    // CC sends permission_mode; VS Code sends timestamp
    if (str("permission_mode") !== undefined) {
        return {
            platform: "claude-code",
            sessionId: str("session_id") ?? "",
            transcriptPath: str("transcript_path") ?? "",
            cwd: str("cwd") ?? "",
            permissionMode: str("permission_mode") ?? "default",
            toolUseId: str("tool_use_id") ?? "",
        };
    }
    if (str("timestamp") !== undefined) {
        return {
            platform: "vscode",
            timestamp: str("timestamp") ?? "",
            sessionId: str("session_id"),
            transcriptPath: str("transcript_path"),
            cwd: str("cwd"),
            toolUseId: str("tool_use_id") ?? "",
            toolResponse: str("tool_response"),
        };
    }
    return { platform: "unknown" };
}
function parseEvent(raw) {
    const rawTool = raw.tool_name ?? "";
    const tool = normalizeToolName(rawTool);
    const input = raw.tool_input ?? {};
    const cmd = input["command"];
    const command = typeof cmd === "string" ? cmd : undefined;
    const event = raw.hook_event_name ?? "unknown";
    const context = detectContext(raw);
    return { event, tool, rawTool, input, command, context };
}
function shouldFire(event, toolFilter) {
    if (!toolFilter)
        return true;
    return event.tool === toolFilter;
}
function writeDeny(reason) {
    const output = {
        hookSpecificOutput: {
            hookEventName: "PreToolUse",
            permissionDecision: "deny",
            permissionDecisionReason: reason,
        },
    };
    process.stdout.write(JSON.stringify(output));
}
function writeAsk(reason) {
    const output = {
        hookSpecificOutput: {
            hookEventName: "PreToolUse",
            permissionDecision: "ask",
            permissionDecisionReason: reason,
        },
    };
    process.stdout.write(JSON.stringify(output));
}
// --- Public API ---
export function createPolicy(options) {
    const raw = readInput();
    const event = parseEvent(raw);
    if (!shouldFire(event, options.tool)) {
        process.exit(0);
    }
    const result = options.check(event);
    if (!result) {
        process.exit(0);
    }
    if ("deny" in result) {
        writeDeny(result.deny);
    }
    else if ("ask" in result) {
        writeAsk(result.ask);
    }
}
export function createObserver(options) {
    const raw = readInput();
    const event = parseEvent(raw);
    if (!shouldFire(event, options.tool)) {
        process.exit(0);
    }
    options.handle(event);
}
export function createSideEffect(options) {
    const raw = readInput();
    const event = parseEvent(raw);
    if (!shouldFire(event, options.tool)) {
        process.exit(0);
    }
    options.handle(event);
}
// --- Utilities for hook scripts ---
export function appendLog(logDir, filename, data) {
    mkdirSync(logDir, { recursive: true });
    const entry = {
        logged_at: new Date().toISOString(),
        ...(typeof data === "object" && data !== null ? data : { value: data }),
    };
    appendFileSync(`${logDir}/${filename}`, JSON.stringify(entry) + "\n");
}
